Gradients: Enable adding custom gradient when gradients are disabled

andrewserong opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Fixes #36640

This PR adds the ability to set a custom gradient when no gradients are provided by the theme (and the `defaultGradients` setting is switched off). It turns out the bug was due to the `MultipleOrigin` component expecting there to be at least one gradient, so the fix is to check that at least one gradient exists before using that component, otherwise fallback to `SingleOrigin`.

Note: I noticed some outstanding issues with the gradient picker, raised in: https://github.com/WordPress/gutenberg/issues/36899 — the console error thrown when clearing a gradient exists on trunk, and doesn't appear to affect the user experience.  I figure we can look at fixing that separately.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

With a blocks theme active, update the following settings in the `theme.json`:

* `settings.color.defaultGradients` should be set to `false`.
* `settings.color.gradients` should be set to an empty array.

Open up the post editor and add a Group block. Go to set the background color, and select Gradient — you should be able to add a custom gradient as in the screengrab below.

I've also added a Storybook example to make sure it's easy to quickly glance at the component without having to make the above changes to a `theme.json` file. To see the Storybook example:

1. Run `npm run storybook:dev`
2. Go to `http://localhost:50240/?path=/story/components-gradientpicker--with-no-existing-gradients` and check that the custom gradient is displayed and usable.

## Screenshots <!-- if applicable -->

![Kapture 2021-11-26 at 16 47 07](https://user-images.githubusercontent.com/14988353/143533542-ee871391-8112-42fe-85a1-cbf3b63506f2.gif)

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

andrewserong committed:
Gradients: Enable adding custom gradient when gradients are disabled

andrewserong committed:
Simplify fix by moving check to where we determine multiple origins, add storybook and changelog entries

ciampo reviewed (commented):
Hey @andrewserong, I'm not sure I'm following your instructions correctly when testing these changes.

In Storybook the gradient input disappears (leaving only the "clear" button) when toggling the "Disable Custom Gradients" to `true` — I couldn't understand if this behaviour is expected.


https://user-images.githubusercontent.com/1083581/144228908-f4d79b62-a747-4bf5-ae6e-4a005c43b46f.mp4


I also wasn't able to follow your instructions in the block editor — I can't see the "Solid | Gradient" toggle, and therefore I can't continue with the testing.

https://user-images.githubusercontent.com/1083581/144224760-3bae4e90-ae36-46a1-beae-783b99fba200.mp4


Apologies in advance in case I'm missing something on my end!

andrewserong committed:
Fix issue with custom and customGradient settings not being loaded in ColorEdit

andrewserong committed:
Hide clear button if there are no gradients and the custom gradient is disabled

andrewserong commented:
Thanks for testing @ciampo! In my testing it turns out my settings were slightly different so I wound up not seeing the error. But from re-testing with my own test instructions (😅) I could replicate the issues. I believe I've fixed them up now:

### Custom gradient not being available at all

It turns out that the setting for `customGradient` wasn't being loaded in the `ColorEdit` component in `packages/block-editor/src/hooks/color.js`. I _think_ this was due to a change in https://github.com/WordPress/gutenberg/pull/36841 that moved away from checking the `color.custom` and `color.customGradient` settings explicitly, and instead uses `useSetting( 'color' )` — unfortunately `useSetting` has special handling for the custom and customGradient paths, so I think that introduced an error. I've restored getting these settings explicitly, but I'll just CC: @jorgefilipecosta in case reverting this introduces any issues.

### Stray clear button being rendered in Storybook

The added Storybook example is mostly there to show that when the component is rendered with `__experimentalHasMultipleOrigins` and `disableCustomGradients` is set to `false`, that the component still renders (mimicking how the component is used in the color support). So, I don't _think_ the stray `Clear` button would show up in real world usage, but to be safe, I've added an extra check to ensure it doesn't render.

---

Thanks again for testing, and let me know if you find any other issues! (The PHP Unit tests are currently failing, but that seems to be happening across a bunch of PRs right now, so most likely unrelated).

ciampo reviewed (approved):
Thank you for the eplxnation @andrewserong !

Code changes on the Gradient Picker side LGTM, and I can test those changes correctly following the instructions! 🚀 

I'll let @jorgefilipecosta give a final seal of approval regarding the changes in the block editor to how color-settings are parsed.

oandregal commented on packages/block-editor/src/hooks/color.js:
```diff
@@ -219,13 +219,16 @@ export function ColorEdit( props ) {
 		link: isLinkEnabled,
 	} = useSetting( 'color' ) || {};
 
+	// The following color settings need to be accessed explicitly due to
```
Oh, that's a good one. It sounds like a follow-up should be to cover for this case in `useSetting` (so the deprecated flags are still considered when accessing the top-level keys `color`, `typography`, `spacing`).

oandregal replied:
https://github.com/WordPress/gutenberg/issues/37094

FYI @jorgefilipecosta @youknowriad

oandregal replied:
While it can be a bit more rare given that themes provide the settings via `theme.json` or we retrofit them, would be worth doing the same for `color.palette` and `color.gradients` (accessing them directly here) just to be sure?

youknowriad replied:
Should we deprecate useSetting with top level keys? `color`, `typography`, `spacing`. I feel this was never considered to be a valid usage.

If we want to keep them, we'd need to "normalize" the settings before running the useSetting logic.

oandregal replied:
I've reverted accessing `color` directly for all the changes I've found https://github.com/WordPress/gutenberg/pull/36900/commits/6a825e63e23bc65ae98983a5f2061a62e6ff00e6

I know this wasn't directly related to custom gradients but it's the source of the bug, so it's best addressed together. It's also one less PR to backport.

_(Andrew, I hope you don't mind I have pushed directly to this PR. I did it so timezones work in our favor, not against)._

andrewserong replied:
> (Andrew, I hope you don't mind I have pushed directly to this PR. I did it so timezones work in our favor, not against).

@oandregal not at all! Thanks so much for picking up this PR, it's always good to get timezones to work in our favor, much appreciated 🙇

oandregal committed:
Revert to access color props specifically

oandregal commented:
For the record, I see `Block_Fixture_Test::test_kses_doesnt_change_fixtures` is failing, but it's also failing in `trunk` and it's unrelated to this PR.

oandregal commented:
I'm going to merge this one, the failing test needs to be investigated independently.

youknowriad commented:
The failing test was actually a valid failure from this PR, I'm trying to fix it now.

jorgefilipecosta commented:
Hi @andrewserong thank you for fixing the issue. On https://github.com/WordPress/gutenberg/pull/36841/. I missed the fact that we need the back-compatibility on some useSetting('color....') settings. And we can not do useSetting( 'color' ).

> It turns out that the setting for customGradient wasn't being loaded in the ColorEdit component in packages/block-editor/src/hooks/color.js. I think this was due to a change in #36841 that moved away from checking the color.custom and color.customGradient settings explicitly, and instead uses useSetting( 'color' ) — unfortunately useSetting has special handling for the custom and customGradient paths, so I think that introduced an error. I've restored getting these settings explicitly, but I'll just CC: @jorgefilipecosta in case reverting this introduces any issues.

The change brings back the bug of colors only being applied by slug on the highest priority origin. But PR https://github.com/WordPress/gutenberg/pull/37108 is fixing this.
